javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -275,5 +283,358 @@ object Matcher {
+    // to the field matched of each node.
+    // If a group did not match, stringForGroupNumber can return null
+    // This is consistent with JVM regexs
+    def setMatch(stringForGroupNumber: Int => String): Unit = {
+      this.matched = stringForGroupNumber(newGroup)
+      nodeType.fold(
+          r => (),
+          children => children.foreach(_.setMatch(stringForGroupNumber))
+      )
+    }
+
+    // Propagates the start position of this node to its descendants
+    def propagateStart(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldLeft(start) {
+          case (newStart, child) => child.setStartReturnEnd(newStart)
+        }
+      )
+    }
+
+    // Propagates the end position of this node to its descendants
+    def propagateEnd(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldRight(end) {
+          case (child, newEnd) => child.setEndReturnStart(newEnd)
+        }
+      )
+    }
+
+    def setEndReturnStart(newEnd: Int): Int = {
+      start = originalGroup match {
+        case OriginallyWrapped.Absolute() => newEnd
+        case _ =>
+          if (matched == null) -1
+          else newEnd - matched.length
+      }
+      propagateStart()
+      start
+    }
+    def setStartReturnEnd(offset: Int): Int = {
+      start = if (matched == null) -1
+              else offset
+
+      originalGroup match {
+        case OriginallyWrapped.Repeater(_) => propagateEnd()
+        case _ =>                          propagateStart()
+      }
+      originalGroup match {
+        case OriginallyWrapped.Absolute() => offset
+        case _ => end
+      }
+    }
+
+    def end: Int = start + (if (matched == null) 0 else matched.length)
+
+    def buildRegex: String =
+      originalGroup.fold(groupNum => "(", (left: String, right: String) => "((?:" + left) +
+      nodeType.     fold(regex => regex,
+        children => "(?:" + children.map(_.buildRegex).mkString("") + ")") +
+      originalGroup.fold(groupNum => ")", (left: String, right: String) => ")" + right + ")")
+
+    def obtainMapping: Map[Int, Node] = {
+       originalGroup.fold(groupNum => Map(groupNum -> this), (_, _) => Map[Int, Node]()) ++
+       nodeType.     fold(regex => Map[Int, Node](), children =>
+         children.foldLeft(Map[Int, Node]()){
+           case (mapping, child) => mapping ++ child.obtainMapping
+         })
+    }
+
+    def simplify: Node = {
+      this match {
+        case Node(OriginallyGroupped(nextGroupIndex),
+                  ParentNode(Seq(Node(OriginallyWrapped("", ""),
+                                      RegexLeaf(regex))))) =>
+          Node(OriginallyGroupped(nextGroupIndex), RegexLeaf(regex))
+        case _ => this
+      }
+
+    }
+  }
+
+  // Wraps every char in the regexp in a group to find the intermediate positions.
+  // Same for subgroups.
+  // (?:AB)C => (?:(AB))(?:(C))
```

Comment:
I always insert the `(?: )` to make sure the repeating modifiers apply on the entire regular expression. Indeed, consider the following regex:

    ab*c

This one I transform it to

    (a)((?:b)*)(c)

so that the `(?:` operator is transparent., and it is not different if there are multiple chars under the star.

I fixed the doc. Originally, I was matching every single char. But later I realised that I could simplify the regex by merging chars if they were not part of different original groups.